src/libxml/errors.cxx

```diff
@@ -117,6 +120,96 @@ namespace impl
 // desirable). So we collect the errors and "replay" them after returning from
 // C code.
 
+std::string build_message(const std::string& what, const xmlError& error)
+{
+    std::ostringstream oss;
+
+    // We currently don't decode domain and code to their symbolic
+    // representation as it doesn't seem to be worth it in practice, the error
+    // message is usually clear enough, while these numbers can be used for
+    // automatic classification of messages.
+    oss << "XML " << what << " " << error.domain << "." << error.code << ": ";
+    if (error.message && *error.message)
+    {
+        if (error.message[std::strlen(error.message) - 1] == '\n')
+            error.message[std::strlen(error.message) - 1] = '\0';
+        oss << error.message;
+    }
+
+    if (error.file)
+    {
+        oss << " at " << error.file;
+        if (error.line)
+        {
+            oss << ":" << error.line;
+
+            // Column information, if available, is passed in the second int
+            // field (first one is used with the first string field, see below).
+            if (error.int2)
+            {
+                oss << "," << error.int2;
+            }
+        }
+    }
+
+    if (error.str1)
+    {
+        oss << " while processing \"" << error.str1 << "\"";
+        if (error.int1)
+        {
+            oss << " at position " << error.int1;
+        }
+    }
+
+    return oss.str();
+}
+
+extern "C" void cb_messages_structured_error(void *out, xmlErrorPtr error)
+{
+    try
+    {
+        // This is not supposed to happen, but don't crash if it ever does.
+        if (!error)
+            return;
+
+        error_messages* const messages = static_cast<error_messages*>(out);
+
+        switch (error->level)
+        {
+            case XML_ERR_NONE:
+                break;
+
+            case XML_ERR_WARNING:
+                // Some libxml warnings are pretty fatal errors, e.g. failing
+                // to open the input file is reported as a warning with code
+                // XML_IO_LOAD_ERROR. Consider all IO warnings as fatal errors
+                // to work around this.
+                if (error->domain != XML_FROM_IO)
```

Comment:
Isn't this what is causing the external DTD error (which arguably really should be a warning, as in `xmllint`)?